auth.php

```diff
@@ -1,128 +1,198 @@
-                return false;
-            }
+    /**
+     * Overwrite authplain cleaning
+     *
+     * @param string $user
+     * @return string
+     */
+    public function cleanUser($user)
+    {
+        return strtolower($user);
+    }
 
-            $query = array();
-            parse_str(base64_decode($sso), $query);
+    /**
+     * Store login data in session
+     *
+     * @param string $user
+     * @param array $userinfo
+     * @return true
+     */
+    protected function storeLogin($user, $userinfo)
+    {
+        global $USERINFO;
+        global $INPUT;
 
-            // verify nonce with generated nonce
-            if ($query['nonce'] !== $this->prev_nonce) {
-                msg($this->getLang('sso_failed'), -1);
-                return false;
-            }
+        $USERINFO = $userinfo;
+        $INPUT->server->set('REMOTE_USER', $user);
 
-            msg($this->getLang('sso_success'), 1);
+        $_SESSION[DOKU_COOKIE]['auth']['user'] = $user;
+        $_SESSION[DOKU_COOKIE]['auth']['info'] = $userinfo;
 
-            // login user
-            $_SESSION['authdiscourse_login'] = $query;
-            return true;
-        }
-        return false;
+        return true;
     }
 
-    // Returns the external SSO login URL.
-    public function getLoginURL() {
-        if (empty($this->login_url))
-            $this->login_url =  $this->generateLoginURL();
-        return $this->login_url;
+    /**
+     * Creates a new user
+     *
+     * @param array $userinfo
+     * @return bool|int|null
+     */
+    protected function addUser($userinfo)
+    {
+        $pwd = auth_pwgen($userinfo['user']);
+
+        return $this->triggerUserMod(
+            'create',
+            [
+                $userinfo['user'],
+                $pwd,
+                $userinfo['name'],
+                $userinfo['mail'],
+                $userinfo['grps'],
+            ]
+        );
     }
 
-    // Generates a URL to the SSO server.
-    private function generateLoginURL() {
-        $payload =  base64_encode(http_build_query(array(
-            'nonce' => $this->nonce,
-            'return_sso_url' => "https://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]",
-        )));
-        $request = array(
-            'sso' => $payload,
-            'sig' => hash_hmac('sha256', $payload, $this->sso_secret),
-        );
-        return $this->sso_url.'?'.http_build_query($request);
+    /**
+     * Validate SSO token/nonce and signature
+     *
+     * @param string $sig
+     * @param string $sso
+     * @return bool
+     */
+    protected function validateToken($sig, $sso)
+    {
+        if (!isset($_COOKIE[self::TOKEN_COOKIE])) {
+            return false;
+        }
+
+        $sso = urldecode($sso);
+        $query = [];
+        parse_str(base64_decode($sso), $query);
+
+        $comp = hash_hmac('sha256', $sso, $this->getConf('sso_secret'));
+
+        return $comp === $sig && $query['nonce'] === $_COOKIE[self::TOKEN_COOKIE];
```

Comment:
The HMAC comparison should be constant time (`hash_equals`?).

Reply:
You're right. I fixed it.